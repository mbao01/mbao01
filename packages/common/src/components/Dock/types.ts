import type { LucideProps } from "lucide-react";
import React from "react";
import { type VariantProps } from "../../libs";
import {
  getDockButtonClasses,
  getDockClasses,
  getDockIconClasses,
  getDockLabelClasses,
} from "./constants";

export type DockProps = React.HTMLAttributes<HTMLDivElement> & VariantProps<typeof getDockClasses>;

export type DockButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof getDockButtonClasses>;

export type DockLabelProps = React.HTMLAttributes<HTMLSpanElement> &
  VariantProps<typeof getDockLabelClasses>;

type P = Omit<LucideProps, "ref"> & React.RefAttributes<SVGSVGElement>;
export type DockIconProps = P &
  VariantProps<typeof getDockIconClasses> & {
    icon: React.ForwardRefExoticComponent<P>;
  };
