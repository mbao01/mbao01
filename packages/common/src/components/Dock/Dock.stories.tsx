import type { Meta, StoryObj } from "@storybook/react";
import { CogIcon, HomeIcon, InboxIcon } from "lucide-react";
import { Dock } from "./Dock";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Dock,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    docs: {
      description: {
        component:
          '⚠️ `<meta name="viewport" content="viewport-fit=cover">is required for responsiveness of the dock in iOS`',
      },
    },
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs", "skip-test"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Dock>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    className: "w-96 mx-auto",
  },
  render: ({ size, ...args }) => {
    return (
      <Dock size={size} {...args}>
        <Dock.Button>
          <Dock.Icon icon={HomeIcon} size={size} />
          <Dock.Label>Home</Dock.Label>
        </Dock.Button>

        <Dock.Button active>
          <Dock.Icon icon={InboxIcon} size={size} />
          <Dock.Label>Inbox</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={CogIcon} size={size} />
          <Dock.Label>Setting</Dock.Label>
        </Dock.Button>
      </Dock>
    );
  },
};

export const TinyDock: Story = {
  args: {
    size: "xs",
    className: "w-96 mx-auto",
  },
  render: ({ size, ...args }) => {
    return (
      <Dock size={size} {...args}>
        <Dock.Button>
          <Dock.Icon icon={HomeIcon} size={size} />
          <Dock.Label>Home</Dock.Label>
        </Dock.Button>

        <Dock.Button active>
          <Dock.Icon icon={InboxIcon} size={size} />
          <Dock.Label>Inbox</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={CogIcon} size={size} />
          <Dock.Label>Setting</Dock.Label>
        </Dock.Button>
      </Dock>
    );
  },
};

export const PrimaryDock: Story = {
  args: {
    variant: "primary",
    className: "w-96 mx-auto",
  },
  render: ({ size, ...args }) => {
    return (
      <Dock size={size} {...args}>
        <Dock.Button active>
          <Dock.Icon icon={HomeIcon} size={size} />
          <Dock.Label>Home</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={InboxIcon} size={size} />
          <Dock.Label>Inbox</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={CogIcon} size={size} />
          <Dock.Label>Setting</Dock.Label>
        </Dock.Button>
      </Dock>
    );
  },
};

export const OutlineDock: Story = {
  args: {
    outline: true,
    variant: "primary",
    className: "w-96 mx-auto",
  },
  render: ({ size, ...args }) => {
    return (
      <Dock size={size} {...args}>
        <Dock.Button>
          <Dock.Icon icon={HomeIcon} size={size} />
          <Dock.Label>Home</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={InboxIcon} size={size} />
          <Dock.Label>Inbox</Dock.Label>
        </Dock.Button>

        <Dock.Button active>
          <Dock.Icon icon={CogIcon} size={size} />
          <Dock.Label>Setting</Dock.Label>
        </Dock.Button>
      </Dock>
    );
  },
};
