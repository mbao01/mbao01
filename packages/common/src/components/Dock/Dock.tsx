import type { DockButtonProps, DockIconProps, DockLabelProps, DockProps } from "./types";
import { cn } from "../../utilities";
import {
  getDockButtonClasses,
  getDockClasses,
  getDockIconClasses,
  getDockLabelClasses,
} from "./constants";

const Dock = ({ size, variant, outline, children, className, ...props }: DockProps) => (
  <div className={cn(getDockClasses({ size, variant, outline }), className)} {...props}>
    {children}
  </div>
);

const DockButton = ({ active, children, className, ...props }: DockButtonProps) => (
  <button className={cn(getDockButtonClasses({ active }), className)} {...props}>
    {children}
  </button>
);

const DockLabel = ({ className, children, ...props }: DockLabelProps) => (
  <span className={cn(getDockLabelClasses(), className)} {...props}>
    {children}
  </span>
);

const DockIcon = ({ icon, size, className, children, ...props }: DockIconProps) => {
  const Icon = icon;
  return (
    <Icon className={cn(getDockIconClasses({ size }), className)} {...props}>
      {children}
    </Icon>
  );
};

Dock.Button = DockButton;
Dock.Label = DockLabel;
Dock.Icon = DockIcon;

export { Dock };
