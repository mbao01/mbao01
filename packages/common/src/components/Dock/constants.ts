import { cva } from "../../libs";

export const getDockClasses = cva("dock", {
  variants: {
    size: {
      xs: "dock-xs",
      sm: "dock-sm",
      md: "dock-md",
      lg: "dock-lg",
      xl: "dock-xl",
    },
    variant: {
      accent: "bg-accent text-accent-content border-accent",
      default: "bg-default text-default-content border-default",
      error: "bg-error text-error-content border-error",
      ghost: "bg-ghost text-ghost-content border-transparent",
      info: "bg-info text-info-content border-info",
      neutral: "bg-neutral text-neutral-content border-neutral",
      primary: "bg-primary text-primary-content border-primary",
      secondary: "bg-secondary text-secondary-content border-secondary",
      success: "bg-success text-success-content border-success",
      warning: "bg-warning text-warning-content border-warning",
    },
    outline: {
      true: "border bg-default",
    },
  },
  compoundVariants: [
    { outline: true, variant: "accent", className: "text-accent" },
    { outline: true, variant: "default", className: "text-default" },
    { outline: true, variant: "error", className: "text-error" },
    { outline: true, variant: "ghost", className: "text-ghost" },
    { outline: true, variant: "info", className: "text-info" },
    { outline: true, variant: "neutral", className: "text-neutral" },
    { outline: true, variant: "primary", className: "text-primary" },
    { outline: true, variant: "secondary", className: "text-secondary" },
    { outline: true, variant: "success", className: "text-success" },
    { outline: true, variant: "warning", className: "text-warning" },
  ],
});

export const getDockButtonClasses = cva("", {
  variants: {
    active: {
      true: "dock-active",
    },
  },
});

export const getDockIconClasses = cva("", {
  variants: {
    size: {
      xs: "size-3",
      sm: "size-4",
      md: "size-5",
      lg: "size-6",
      xl: "size-6",
    },
  },
});

export const getDockLabelClasses = cva("dock-label");
