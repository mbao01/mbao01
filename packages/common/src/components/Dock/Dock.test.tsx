import { render, screen } from "@testing-library/react";
import { CogIcon, HomeIcon, InboxIcon } from "lucide-react";
import { Dock } from "./Dock";

describe("Dock", () => {
  it("renders a basic dock", () => {
    const { asFragment } = render(
      <Dock>
        <Dock.Button active>
          <Dock.Icon icon={HomeIcon} />
          <Dock.Label>Home</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={InboxIcon} />
          <Dock.Label>Inbox</Dock.Label>
        </Dock.Button>

        <Dock.Button>
          <Dock.Icon icon={CogIcon} />
          <Dock.Label>Setting</Dock.Label>
        </Dock.Button>
      </Dock>
    );

    const homeButton = screen.getByRole("button", { name: "Home" });
    const inboxButton = screen.getByRole("button", { name: "Inbox" });
    const settingButton = screen.getByRole("button", { name: "Setting" });

    expect(homeButton).toBeVisible();
    expect(inboxButton).toBeVisible();
    expect(settingButton).toBeVisible();

    expect(asFragment()).toMatchSnapshot();
  });
});
