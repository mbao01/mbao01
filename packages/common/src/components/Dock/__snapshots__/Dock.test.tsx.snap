// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dock > renders a basic dock 1`] = `
<DocumentFragment>
  <div
    class="dock"
  >
    <button
      class="dock-active"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-house"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
        />
        <path
          d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
        />
      </svg>
      <span
        class="dock-label"
      >
        Home
      </span>
    </button>
    <button
      class=""
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-inbox"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="22 12 16 12 14 15 10 15 8 12 2 12"
        />
        <path
          d="M5.45 5.11 2 12v6a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2v-6l-3.45-6.89A2 2 0 0 0 16.76 4H7.24a2 2 0 0 0-1.79 1.11z"
        />
      </svg>
      <span
        class="dock-label"
      >
        Inbox
      </span>
    </button>
    <button
      class=""
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-cog"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 20a8 8 0 1 0 0-16 8 8 0 0 0 0 16Z"
        />
        <path
          d="M12 14a2 2 0 1 0 0-4 2 2 0 0 0 0 4Z"
        />
        <path
          d="M12 2v2"
        />
        <path
          d="M12 22v-2"
        />
        <path
          d="m17 20.66-1-1.73"
        />
        <path
          d="M11 10.27 7 3.34"
        />
        <path
          d="m20.66 17-1.73-1"
        />
        <path
          d="m3.34 7 1.73 1"
        />
        <path
          d="M14 12h8"
        />
        <path
          d="M2 12h2"
        />
        <path
          d="m20.66 7-1.73 1"
        />
        <path
          d="m3.34 17 1.73-1"
        />
        <path
          d="m17 3.34-1 1.73"
        />
        <path
          d="m11 13.73-4 6.93"
        />
      </svg>
      <span
        class="dock-label"
      >
        Setting
      </span>
    </button>
  </div>
</DocumentFragment>
`;
